File: src/tf/transform_listener.cpp
#include "torch-tf.h"
#include <tf/transform_listener.h>

TFIMP(tf::TransformListener *, TransformListener, new)()
{
  return new tf::TransformListener();
}

TFIMP(void, TransformListener, delete)(tf::TransformListener *self)
{
  delete self;
}

TFIMP(void, TransformListener, clear)(tf::TransformListener *self)
{
  self->clear();
}

TFIMP(void, TransformListener, getFrameStrings)(tf::TransformListener *self, std::vector<std::string> *result)
{
  self->getFrameStrings(*result);
}

TFIMP(void, TransformListener, lookupTransform)(tf::TransformListener *self, const char *target_frame, const char *source_frame, ros::Time *time, tf::StampedTransform *result)
{
  self->lookupTransform(target_frame, source_frame, *time, *result);
}

TFIMP(bool, TransformListener, waitForTransform)(tf::TransformListener *self, const char *target_frame, const char *source_frame, ros::Time *time, ros::Duration *timeout, std::string *error_msg)
{
  return self->waitForTransform(target_frame, source_frame, *time, *timeout, ros::Duration(0.01), error_msg);
}

TFIMP(bool, TransformListener, canTransform)(tf::TransformListener *self, const char *target_frame, const char *source_frame, ros::Time *time, std::string *error_msg)
{
  return self->canTransform(target_frame, source_frame, *time, error_msg);
}

TFIMP(void, TransformListener, resolve)(tf::TransformListener *self, const char *frame_name, std::string *result)
{
  *result = self->resolve(frame_name);
}

File: TransformListener.lua
local ffi = require 'ffi'
local torch = require 'torch'
local moveit = require 'moveit.env'
local utils = require 'moveit.utils'
local std = moveit.std
local tf = moveit.tf

local TransformListener = torch.class('tf.TransformListener', tf)

local f

function init()
  local TransformListener_method_names = {
    "new",
    "delete",
    "clear",
    "getFrameStrings",
    "lookupTransform",
    "waitForTransform",
    "canTransform",
    "resolve"
  }
  
  f = utils.create_method_table("tf_TransformListener_", TransformListener_method_names)
end

init()

function TransformListener:__init()
  self.o = f.new()
end

function TransformListener:clear()
  f.clear(self.o)
end

function TransformListener:getFrameStrings()
  local result = std.StringVector()
  f.getFrameStrings(self.o, result:cdata())
  return result
end

function TransformListener:lookupTransform(target_frame, source_frame, time, result)
  result = result or tf.StampedTransform()
  f.lookupTransform(self.o, target_frame, source_frame, time:cdata(), result:cdata())
  return result
end

function TransformListener:waitForTransform(target_frame, source_frame, time, timeout)
  local error_msg = std.String()
  local success = f.waitForTransform(self.o, target_frame, source_frame, time:cdata(), timeout:cdata(), error_msg:cdata())
  return success, error_msg:get()
end

function TransformListener:canTransform(target_frame, source_frame, time)
  local error_msg = std.String()
  local success = f.canTransform(self.o, target_frame, source_frame, time:cdata(), error_msg:cdata())
  return success, error_msg:get()
end

function TransformListener:resolve(frame_name)
  local name = std.String()
  f.resolve(self.o, frame_name, name:cdata())
  return name:get()
end
